resolver/dns/go19.go

```diff
@@ -0,0 +1,43 @@
+// +build go1.9
+
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package dns
+
+import (
+	"net"
+	"strings"
+
+	"golang.org/x/net/context"
+)
+
+var customAuthorityResolver = func(authority string) (netResolver, error) {
+	if _, _, err := net.SplitHostPort(authority); err != nil && strings.Contains(err.Error(), "missing port in address") {
+		authority = authority + ":53"
+	}
+
+	return &go18upResolver{
+		resolver: &net.Resolver{
+			PreferGo: true,
+			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
```

Comment:
Instead of inline the function definition here, we define a global variable to be the same as the inline function, and assign the global variable to `Dial` here.

And in the test, we substitute the global variable value to a fake dial function which does not dial to a dns server, but instead verifying the dial target is correct.

Reply:
Aha. Gotcha. Will fix later.

Reply:
The dial function needs the `authority` var from customAuthorityResolver, so i cant really move it to a global. 

I'd end up with something like this, which seems like a lot of code for no benefit.
```golang
var customAuthorityDialler = func(authority string) func(ctx context.Context, network, address string) (net.Conn, error) {
	return func(ctx context.Context, network, address string) (net.Conn, error) {
		dialer := net.Dialer{}
		return dialer.DialContext(ctx, network, authority)
	}
}

var customAuthorityResolver = func(authority string) (netResolver, error) {
	if _, _, err := net.SplitHostPort(authority); err != nil && strings.Contains(err.Error(), "missing port in address") {
		authority = authority + ":53"
	}

	return &go18upResolver{
		resolver: &net.Resolver{
			PreferGo: true,
			Dial:     customAuthorityDialler(authority),
		}}, nil
}
```

What about this instead.. Bare minimum test.

```golang
func TestCustomAuthority(t *testing.T) {
	defer leakcheck.Check(t)

	customAuthority := "4.3.2.1"

	oldCustomAuthorityResolver := customAuthorityResolver
	customAuthorityResolver = func(authority string) (netResolver, error) {
		if authority != customAuthority {
			t.Errorf("wrong custom authority passed to resolver. expected: %s actual: %s", customAuthority, authority)
		}
		return nil, errors.New("not implemented")
	}
	defer func() {
		customAuthorityResolver = oldCustomAuthorityResolver
	}()

	b := NewBuilder()
	cc := &testClientConn{target: "foo.bar.com"}
	b.Build(resolver.Target{Endpoint: "foo.bar.com", Authority: customAuthority}, cc, resolver.BuildOption{})

}
```

Reply:
The bare minimum test cannot test the logic(e.g appending ":53" for authority without port) inside `customAuthorityResolver`. We need something like your first code snippet.